Don't cancel aborted jobs immediately

hubertp opened the pull request:
### Pull Request Description

Rather than cancelling Futures that capture jobs' logic, this change introduces a two-level system:
- interrupt all jobs softly via ThreadInterrupted at safepoints
- if safepoint is not executed within some time period or it is but the job is still not cancelled, trigger a hard-interrupt by cancelling the job explicitly, if possible

Closes #11084.

### Important Notes

This change, while effectively does not alter the end result (ending jobs that take too much time), should reduce chances of IO Exceptions. It should also help identify where we lack safepoints polling or what java calls should be improved.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

hubertp committed:
Don't cancel aborted jobs immediately

Rather than cancelling Futures that capture jobs' logic,
this change introduces a two-level system:

- interrupt all jobs softly via ThreadInterrupted at safepoints
- if safepoint is not executed within some time period or it is
  but the job is still not cancelled, trigger a hard-interrupt
  by cancelling the job explicitly, if possible

Closes #11084.

hubertp commented:
Actually, the current solutions is wrong. Marking a Future as a (soft) cancel not only changes the state to `Cancelled` but also prevents further cancellations, potentially with interrupts, to be enacted.

hubertp committed:
Only cancel Future when you mean it

Soft-cancelling a future only to later call it with `mayInterrupt` set
to `true` has no effect in the latter case.
Changed the logic so that interrupting a Future will really enforce it.

Ocassionally some commands should not attempt to run soft cancellations
- we know they will re-execute the program.

hubertp committed:
Merge branch 'develop' into wip/hubert/11084-interrupts

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -78,6 +84,84 @@ final class JobExecutionEngine(
+            val timeout = nextToCancel
+              .map(j => Math.min(j._1 - at, forceInterruptTimeout))
+              .getOrElse(forceInterruptTimeout)
+            Thread.sleep(timeout)
```
No `Thread.sleep` in production code, please.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -35,8 +35,14 @@ final class JobExecutionEngine(
   private val context = interpreterContext.executionService.getContext
 
+  private val pendingCancellationsExecutor =
+    context.newFixedThreadPool(1, "pending-cancellations-timer", false)
```
`EnsoContext.newFixedThreadPool` returns [ExecutorService](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ExecutorService.html). Can we wrap the default executor services with `SafepointAwareExecutorService` and whenever a method returning `Future` return a wrapper `SafepointAwareFuture` that would `cancel(true)` by asking `ThreadManager` first and only later emit an interrupt?

JaroslavTulach reviewed (commented):
I have a feeling that by implementing our own `Future` + wrapping `ExecutionService` the code would be simpler.

Would it be possible to implement this without using new thread - e.g. `pendingCancellationsExecutor` at all? Possibly with own `Future.cancel`, right? It would just block until _safepoint_ or later `interrupt()` stops the `Future` computation.

I would feel more confident if the code was next to `EnsoContext`. First of all it would encapsulate all the logic and would be easier to unit test. Next, but not last: It would be written in more easily readable language.

hubertp committed:
Replace Thread.sleep with Future.get

No while loops etc, it's much easier to reason about what is soft and
hard interrupt supposed to do.

hubertp committed:
Better comments/logs

hubertp commented:
I got rid of `Thread.sleep` and removed the most controversial bits. I agree that would experiment with a more safepoint-aware execution in futures but I'd rather avoid implementing it in this PR. I created a separate ticket so that I don't forget our intentions for future enhancement.

hubertp committed:
Merge branch 'develop' into wip/hubert/11084-interrupts

hubertp committed:
nit

hubertp committed:
Merge branch 'develop' into wip/hubert/11084-interrupts

4e6 reviewed (approved):
> This change, while effectively does not alter the end result (ending jobs that take too much time), should reduce chances of IO Exceptions. It should also help identify where we lack safepoints polling or what java calls should be improved.

Ok, this makes sense

JaroslavTulach commented on engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala:
```diff
@@ -230,7 +230,7 @@ class RuntimeAsyncCommandsTest
     while (!isProgramStarted && iteration < 100) {
       val out = context.consumeOut
-      Thread.sleep(200)
+      Thread.sleep(100)
```
Common multithreaded pattern for awaiting output would be:
- modify `InterpreterContext._output` to have a subclass of `ByteArrayOutputStream` that notifies a monitor whenever something is written into it
- add `context.awaitOut(timeOut)` to check the output and `Object.wait(timeOut)` if the desired text hasn't yet arrived
- continue the test immediately when the desired text appears in the log by being notified as soon as possible

hubertp replied:
I left out the logic as it was defined before. We can probably improve it it becomes flaky.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/Job.scala:
```diff
@@ -27,11 +29,32 @@ abstract class Job[+A](
+    *
+    * @param ctx contains suppliers of services to perform a request
+    */
+  def run(implicit ctx: RuntimeContext): A = {
```
Should the behavior of `hasStated` be guaranteed, then this method has to be `final`.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -78,6 +80,82 @@ final class JobExecutionEngine(
+          try {
+            val now                        = System.currentTimeMillis()
+            val timeSinceRequestedToCancel = now - timeRequestedToCancel
+            assert(timeSinceRequestedToCancel > 0)
```
There is `assertInJVM` that is preferred to use when honoring `-ea` is the goal.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:
```diff
@@ -78,6 +80,82 @@ final class JobExecutionEngine(
+                "Threaddump when timeout is reached while waiting for the job " + runningJob.id + " running in thread " + runningJob.job
+                  .threadNameExecutingJob() + " to cancel:\n"
+              )
+              Thread.getAllStackTraces.entrySet.forEach { entry =>
```
Logging this kind of information may turn very useful.

hubertp committed:
PR review

hubertp committed:
Merge branch 'develop' into wip/hubert/11084-interrupts

hubertp committed:
Make test more robust
